### PR TITLE
FEC-8975 QoS data enhancement

### PR DESCRIPTION
**Pull request by RoyBregman**
### Description of the Changes

Added FRAG_LOAD and MANIFEST_LOAD events

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/engines/html5/html5.js [hunk 1]
```diff
@@ -299,6 +301,24 @@ export default class Html5 extends FakeEventTarget implements IEngine {
     return this._el;
   }
 
+  /**
+   * @returns {number} - the number of video frames dropped since the creation of the associated HTMLVideoElement
+   */
+  getDroppedVideoFrames(): ?number {
+    if (this._droppedFramesWatcher) {
+      return this._droppedFramesWatcher._getDroppedAndDecodedFrames()[0];
```
Comment (yairans) [inline src/engines/html5/html5.js]: `_getDroppedAndDecodedFrames` isa private method

File: src/engines/html5/html5.js [hunk 2]
```diff
@@ -642,7 +644,7 @@ export default class Html5 extends FakeEventTarget implements IEngine {
    * @public
    */
   get buffered(): TimeRanges {
-    return this._el.buffered;
+    return this._mediaSourceAdapter ? this._mediaSourceAdapter.buffered : 0;
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/engines/html5/html5.js]: You are not looking at the latest code. This was removed later

File: src/engines/html5/media-source/base-media-source-adapter.js [hunk 3]
```diff
@@ -264,4 +264,12 @@ export default class BaseMediaSourceAdapter extends FakeEventTarget implements I
   get capabilities(): PKMediaSourceCapabilities {
     return this._capabilities;
   }
+
+  get targetBuffer(): number {
+    return NaN;
+  }
+
+  get buffered(): TimeRanges {
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/engines/html5/media-source/base-media-source-adapter.js]: Same here

File: src/player.js [hunk 4]
```diff
@@ -647,6 +647,26 @@ export default class Player extends FakeEventTarget {
...
+    const videoElement = this.getVideoElement();
+    let statsObject: PKStatsObject = {
+      targetBuffer: NaN,
+      availableBuffer: NaN
+    };
+    if (this._engine) {
+      statsObject.targetBuffer = this._engine.targetBuffer;
+      statsObject.availableBuffer = this._engine.availableBuffer;
+      // consideration of the end of the playback in the target buffer calc
+      if (!this.isLive()) {
```
Comment (OrenMe) [inline src/player.js]: I think you can simplify this to 
```javascript 
if (!this.isLive() && videoElement) {
  statsObject.targetBuffer = Math.min(statsObject.targetBuffer, videoElement.duration - videoElement.currentTime);
}
```
You can also move the `const videoElement = this.getVideoElement();`  inside the of as you only use it there 

Also, did you add the new engine members to flash and YouTube engines?

Comment (OrenMe) [reply src/player.js]: Also, shouldn’t this be `if ((!this.isLive() || this.isDvr()) && vidoeElement)`

File: src/player.js [hunk 5]
```diff
@@ -647,6 +647,26 @@ export default class Player extends FakeEventTarget {
     }
   }
 
+  get stats(): PKStatsObject {
+    const videoElement = this.getVideoElement();
+    let statsObject: PKStatsObject = {
+      targetBuffer: NaN,
+      availableBuffer: NaN
+    };
+    if (this._engine) {
```
Comment (OrenMe) [inline src/player.js]: You can also substitute this with `if (videoElement)` and remove it from the bottom check as `this.getVideoElement()` validates for `_engine` as well so you can omit it here

File: src/player.js (hunk 4 above)
Comment (RoyBregman) [reply src/player.js]: The dvr handling is being done in HLS and Dash as this depends on the player implementation

Comment (RoyBregman) [reply src/player.js]: > I think you can simplify this to
> 
> ```js
> if (!this.isLive() && videoElement) {
>   statsObject.targetBuffer = Math.min(statsObject.targetBuffer, videoElement.duration - videoElement.currentTime);
> }
> ```
> 
> You can also move the `const videoElement = this.getVideoElement();` inside the of as you only use it there
> 
> Also, did you add the new engine members to flash and YouTube engines?

Nope, needs to be done...

File: src/player.js (hunk 5 above)
Comment (RoyBregman) [reply src/player.js]: I know it sounds weird, but can there be an engine but no video elements?

Comment (OrenMe) [reply src/player.js]: No, `this._el` in the engines, which is the response of `getVideoElement` is done in engine constructor.
Also, as mentioned, `this.getVideoElement()` is protected by itself with check fir `_engine`

File: src/player.js [hunk 6]
```diff
@@ -657,10 +656,9 @@ export default class Player extends FakeEventTarget {
       statsObject.targetBuffer = this._engine.targetBuffer;
       statsObject.availableBuffer = this._engine.availableBuffer;
       // consideration of the end of the playback in the target buffer calc
-      if (!this.isLive()) {
-        statsObject.targetBuffer = videoElement
-          ? Math.min(statsObject.targetBuffer, videoElement.duration - videoElement.currentTime)
-          : statsObject.targetBuffer;
+      if (!this.isLive() && this.getVideoElement()) {
+        const videoElement = this.getVideoElement();
+        statsObject.targetBuffer = Math.min(statsObject.targetBuffer, videoElement.duration - videoElement.currentTime);
```
Comment (OrenMe) [inline src/player.js]: ok, I now understand this is not safe, as accessing the videoElement duration and currentTime is not guaranteed, as not all are HTML5 video element.
So I guess best place at the moment is in the base media source adapter in the `targetBuffer` getter

**Pull request merged**